### PR TITLE
Full library functionality

### PR DESCRIPTION
**Pull request by kaciulis**
Implemented the library:

- Cgroup flow, support for cgroup, cgroup2 and hybrid mode
- Forking and reexecuting child process in new namespaces
- A job can be started and stopped
- A job can be streamed by multiple listeners

### CODE REVIEW

File: library/worker.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	j.jobsMapMutex.Lock()
+	defer j.jobsMapMutex.Unlock()
+
+	if _, exists := j.jobs[job.id]; exists {
+		return nil, errors.New("uuid generation overlap")
+	}
+
+	j.jobs[job.id] = job
+
+	return job, nil
```
Comment (zmb3) [inline library/worker.go:77]: It doesn't look like you actually start a job here.

File: library/worker.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	if job.outputStream == nil {
+		return nil, errors.New("no output stream initialized")
+	}
+
+	return job.outputStream.getStream(), nil
+}
+
+// getJob is a thread safe way to get a job by its ID
+func (j *JobWorker) getJob(jobID string) (*Job, error) {
+	if j.jobs == nil {
```
Comment (zmb3) [inline library/worker.go:127]: Under what circumstances could this happen?

File: library/worker.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		return err
+	}
+
+	return job.Stop()
+}
+
+// Status is the exposed implementation of the status
+// query functionality of the library.
+func (j *JobWorker) Status(jobID string) (int, int, error) {
```
Comment (zmb3) [inline library/worker.go:94]: It would be nice to have a separate type for the job status so it's easier to see which return value corresponds to the status and which is the exit code.

**Review by zmb3 — CHANGES_REQUESTED** [review]
First pass - I added a few comments throughout.

The streaming logic for me is a bit hard to follow, and there are no tests or runnable code in this PR that I use to see how things work. Could you please add some tests and/or a simple program that uses this library so that we can try running your code?

File: library/worker.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	jobCommand.Stderr = job.outputStream
+	// With a bidirectional stream from the client to the Stdin, this library could support remote shell functionality
+	jobCommand.Stdin = nil
+
+	j.jobsMapMutex.Lock()
+	defer j.jobsMapMutex.Unlock()
+
+	if _, exists := j.jobs[job.id]; exists {
+		return nil, errors.New("uuid generation overlap")
+	}
```
Comment (russjones) [inline library/worker.go:73]: I don't think this would actually ever happen.

File: library/worker.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+// The listener can set up a channel listening loop where GetBytes()
+// returns a receiver-channel of the streams output and
+func (j *JobWorker) Stream(jobID string) (*streamChannel, error) {
+	job, err := j.getJob(jobID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if job.outputStream == nil {
```
Comment (russjones) [inline library/worker.go:118]: Similar question to what @zmb3 asked here: https://github.com/kaciulis/job-worker-service/pull/2/files#r710313597

Here and also when you are trying to access `status` on a job, you have defensive checks everywhere if a field is not set. 

Any reason why you are not initializing these variables when you create your structs? Would make your code much simpler.

File: library/cgroup.go [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+	}
+
+	f, err := os.Open("/proc/self/mountinfo")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
```
Comment (r0mant) [inline library/cgroup.go:130]: `scanner.Err()` isn't checked.

File: library/stream.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+		select {
+		case output := <-j.byteChannel:
+			j.bufferMux.Lock()
+			streamChannels := j.streamChannels
+			j.bufferMux.Unlock()
+
+			for _, channelObject := range streamChannels {
+				go func(targetStream *streamChannel, bytes []byte) {
+					targetStream.channel <- bytes
+				}(channelObject, output)
```
Comment (russjones) [inline library/stream.go:61]: If the client disconnects in the middle of a stream, would this cause a goroutine leak?

File: library/job.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	 return nil
+}
+
+// Stop is an exposed method used to Interrupt the jobs process.
+func (j *Job) Stop() error {
+	if err := j.cmd.Process.Signal(syscall.SIGINT); err != nil {
+		return err
+	}
+
+	j.setStatus(JobStoppedStatus, j.cmd.ProcessState.ExitCode())
```
Comment (r0mant) [inline library/job.go:71]: What if the job has already finished on its own prior to reaching here?

File: library/worker.go (hunk 5 above)
Comment (r0mant) [inline library/worker.go:118]: Can this happen?

File: library/stream.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+// GetStream Creates and returns a new StreamingChannel with the current output of the stdout and stderr
+func (j *jobOutputStream) getStream() *streamChannel {
+	j.bufferMux.Lock()
+	defer j.bufferMux.Unlock()
+
+	bytesSoFar := j.buffer.Bytes()
+	channelStream := &streamChannel{channel: make(chan []byte), done: make(chan bool)}
+
+	go func(targetStream *streamChannel, bytes []byte, streamOver bool) {
+		channelStream.channel <- bytes
```
Comment (r0mant) [inline library/stream.go:116]: What happens to this goroutine if the client disconnects without attempting to read anything from the channel?

Comment (r0mant) [reply library/stream.go:116]: Also, isn't there a race condition between this goroutine and "broadcastStream"? Can the client receive newer logs before older ones get streamed?

File: library/cgroup.go [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+	UnknownController CGroupController = "unknown"
+	CPUController     CGroupController = CGroupControllerCPU
+	IOController      CGroupController = CGroupControllerIO
+	MemoryController  CGroupController = CGroupControllerMemory
+	IOV2Controller    CGroupController = CGroupV2ControllerIO
+)
+
+// ResourceLimitation represents the interface of the CGroup controllers actual creation functionality.
+type ResourceLimitation interface {
+	addCGroupV1(string) error // A method to create cgroup v1 files
```
[the remark this answers is not in the document]
Comment (kaciulis) [reply library/cgroup.go:51]: Realistically speaking, due to slow adoption rate of cgroupv2 it’s not realistic to expect the library to work based on the challenges requirements with only cgroupv2. The best I can do is prefer cgroupv2. Most machines should be in hybrid mode and in hybrid mode even if all the cgroups of version 1 are unmounted it still doesn’t allow the cgroupv2 mount to control all of the required controllers. The reasoning behind this scope creep is simple in my mind - there are two requirements - 1) “the library should run on a linux amd64 machine” and 2) “the job should have its resources limited”.  So requirement number 1 doesn’t specify a linux distribution which would then allow me to pick one of the Cgroup versions and requirement number 2 doesn’t give me space to not limit the resources. For example - just because the memory controller is not available to cgroupv2 does not mean I should ignore the memory limitation requirements. I might be overthinking tho’..

File: library/cgroup.go [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+	addCGroupV2(string) error // A method to create cgroup v2 files
```
[the remark this answers is not in the document]
Comment (kaciulis) [reply library/cgroup.go:52]: The reasoning is “by accident” :D

File: library/exec.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+const DefaultReexecArgument = "reexec"
+const LinuxSelfReferenceExecutable = "/proc/self/exe"
+
+// Init is the function that looks if the current process is a fork process.
+//
+// Init should be invoked in the main package init() function.
+func Init() {
+	if os.Args[1] == DefaultReexecArgument {
+		child()
+		os.Exit(0)
```
[the remark this answers is not in the document]
Comment (kaciulis) [reply library/exec.go]: Yep, Forgot to remove it

File: library/job.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+	 if err != nil {
+		 j.cleanUp()
+
+		 return err
+	 }
+
+	 go func() {
+		 status := JobFinishedStatus
+		 exitCode := j.cmd.ProcessState.ExitCode()
```
[the remark this answers is not in the document]
Comment (kaciulis) [reply library/job.go]: Just wanted a fallback value, but I guess -1 will also do

File: library/job.go [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+		 if err := j.cmd.Wait(); err != nil {
+			 status = JobFailedStatus
+
+			 if exitError, ok := err.(*exec.ExitError); ok {
+				 exitCode = exitError.ExitCode()
+
+				 if _, err = j.outputStream.Write(exitError.Stderr); err != nil {
+					 exitCode = 255
```
[the remark this answers is not in the document]
Comment (kaciulis) [reply library/job.go]: Late night overthinking - wanted a way to inform the forked process that the child process was unable to write its error message. As you’re a pointing at it - now it seems silly.

File: library/job.go [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+	j.setStatus(JobStoppedStatus, j.cmd.ProcessState.ExitCode())
+	j.cleanUp()
+
+	return nil
+}
+
+// GetStatusAndExitCode is an exposed method that is thread safe. Used to access the current status and exitCode of the process.
+func (j *Job) GetStatusAndExitCode() (int, int, error) {
+	j.mux.Lock()
```
[the remark this answers is not in the document]
Comment (kaciulis) [reply library/job.go]: totally right, changed to RLock/RUnlock.

File: library/stream.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+			j.bufferMux.Unlock()
+
+			// If the stream ends - each streamChannel will be informed
+			for _, channelObject := range streamChannels {
+				go func(targetStream *streamChannel) {
+					targetStream.close()
+				}(channelObject)
+			}
+
+			j.ended = true
```
[the remark this answers is not in the document]
Comment (kaciulis) [reply library/stream.go]: Makes sense, tho’ while I was testing this I wasn’t able to break this part. Either way - you’re right, this should by in a lock/unlock context, fixed

File: library/worker.go (hunk 1 above)
Comment (kaciulis) [reply library/worker.go:77]: The job itself has an exposed method called Run() which starts the job. I’m totally fine with changing it to also run in the Worker.Start() if you think that it is necessary. I Added a couple of example programs for you to check on how that would work.

File: library/worker.go (hunk 2 above)
Comment (kaciulis) [reply library/worker.go:139]: On none, it's just future-proofing

File: library/worker.go (hunk 3 above)
Comment (kaciulis) [reply library/worker.go]: To be honest - have no idea why I didn’t make it in the first place

File: library/worker.go (hunk 4 above)
Comment (kaciulis) [reply library/worker.go:73]: Once in a million years, but still possible :D

File: library/worker.go (hunk 5 above)
Comment (kaciulis) [reply library/worker.go:118]: I am initialising these variables when I’m creating these structs. It’s just a way to foolproof the code for the future. It’s definitely not needed since it’s expecting for the worst case scenarios, where simple tests would suffice instead. Should I remove these paranoid checks?

File: library/stream.go (hunk 7 above)
Comment (kaciulis) [reply library/stream.go:60]: This case feels like it should be handled by the service that uses this library, not the library itself. I specifically made a test program for this case to show to you by what I mean.

File: library/cgroup.go (hunk 6 above)
Comment (kaciulis) [reply library/cgroup.go:130]: Ups

File: library/job.go (hunk 8 above)
Comment (kaciulis) [reply library/job.go:72]: Good point! I think I should check if the status is running and only then change it to stopped.

File: library/worker.go (hunk 5 above)
Comment (kaciulis) [reply library/worker.go:118]: At this moment in time - no

File: library/stream.go (hunk 9 above)
Comment (kaciulis) [reply library/stream.go:117]: Then the goroutine freezes. The service that uses this library should ensure that on a clients disconnect, the streaming will be stopped. Check the example projects to get what I mean.  There would be a race condition if there would be multiple goroutines created at the same time to push bytes to the channel I think. In this case each new byte array is distributed between the streams and the byte buffer with a mutex lock. Also, the writing to the main broadcasting stream happens in a locked manner, where the individual streams should be able to read without any locks.
